### ISSUE TITLE
java.lang.ExceptionInInitializerError Firebase Performance

### ISSUE THREAD

**Issue by zsweigart** [firebase/firebase-android-sdk#225]
* Android Studio version: 3.2.1
  * Firebase Component: com.google.firebase:firebase-auth:16.1.0, com.google.firebase:firebase-database:16.0.5, com.google.firebase:firebase-config:16.1.2, com.google.firebase:firebase-messaging:17.3.4, com.google.firebase:firebase-perf:16.2.3",
  * Component version: see above

```
java.lang.ExceptionInInitializerError
        at com.google.firebase.perf.internal.RemoteConfigManager.a(Unknown Source:17)
        at com.google.firebase.perf.internal.FeatureControl.<init>(Unknown Source:2)
        at com.google.firebase.perf.internal.FeatureControl.a(Unknown Source:10)
        at com.google.firebase.perf.internal.GaugeManager.<init>(Unknown Source:4)
        at com.google.firebase.perf.internal.GaugeManager.a(Unknown Source:22)
        at com.google.android.gms.internal.firebase-perf.s.<init>(Unknown Source:2)
        at com.google.android.gms.internal.firebase-perf.s.a(Unknown Source:1)
        at com.google.firebase.perf.network.FirebasePerfUrlConnection.instrument(Unknown Source:9)
        at com.facebook.GraphRequest.a(SourceFile:1407)
        at com.facebook.GraphRequest.a(SourceFile:1067)
        at com.facebook.GraphRequest.b(SourceFile:1155)
        at com.facebook.GraphRequest.a(SourceFile:1131)
        at com.facebook.GraphRequest.a(SourceFile:1115)
        at com.facebook.GraphRequest.a(SourceFile:1090)
        at com.facebook.GraphRequest.i(SourceFile:984)
        at com.facebook.internal.aa.c(SourceFile:244)
        at com.facebook.internal.aa.b(SourceFile:52)
        at com.facebook.internal.ab.run(SourceFile:135)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalStateException: Default FirebaseApp is not initialized in this process com.coffeemeetsbagel. Make sure to call FirebaseApp.initializeApp(Context) first.
        at com.google.firebase.b.d(SourceFile:240)
        at com.google.firebase.remoteconfig.a.a(Unknown Source:11)
        at com.google.firebase.perf.internal.RemoteConfigManager.<init>(Unknown Source:2)
        at com.google.firebase.perf.internal.RemoteConfigManager.<clinit>(Unknown Source:160)
        at com.google.firebase.perf.internal.RemoteConfigManager.a(Unknown Source:17) 
        at com.google.firebase.perf.internal.FeatureControl.<init>(Unknown Source:2) 
        at com.google.firebase.perf.internal.FeatureControl.a(Unknown Source:10) 
        at com.google.firebase.perf.internal.GaugeManager.<init>(Unknown Source:4) 
        at com.google.firebase.perf.internal.GaugeManager.a(Unknown Source:22) 
        at com.google.android.gms.internal.firebase-perf.s.<init>(Unknown Source:2) 
        at com.google.android.gms.internal.firebase-perf.s.a(Unknown Source:1) 
        at com.google.firebase.perf.network.FirebasePerfUrlConnection.instrument(Unknown Source:9) 
        at com.facebook.GraphRequest.a(SourceFile:1407) 
        at com.facebook.GraphRequest.a(SourceFile:1067) 
        at com.facebook.GraphRequest.b(SourceFile:1155) 
        at com.facebook.GraphRequest.a(SourceFile:1131) 
        at com.facebook.GraphRequest.a(SourceFile:1115) 
        at com.facebook.GraphRequest.a(SourceFile:1090) 
        at com.facebook.GraphRequest.i(SourceFile:984) 
        at com.facebook.internal.aa.c(SourceFile:244) 
        at com.facebook.internal.aa.b(SourceFile:52) 
        at com.facebook.internal.ab.run(SourceFile:135) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
```

#### Steps to reproduce:
Occurs on app launch (sometimes)

#### Relevant Code:

We are manually setting up firebase

```
public class App extends Application {
  public static final String TAG = "App";
  private static App sInstance;

  public static App getInstance() {
    return sInstance;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    sInstance = this;

    setupFirebase();
    FirebasePerformance.startTrace("app startup");
  }

  protected void setupFirebase() {
    FirebaseOptions.Builder builder = new FirebaseOptions.Builder()
        .setApplicationId(getString(R.string.google_app_id))
        .setApiKey(getString(R.string.google_api_key))
        .setDatabaseUrl("URL");
    FirebaseApp.initializeApp(this, builder.build());
  }

  ...
}
```

We have tried the following steps to resolve:
1. Added debug logs to wherever we call facebook graph executeAndWait (per the stack trace) and none of them are hit before this crash
2. Tried upgrading facebook sdk libraries and moving to modules instead of full sdk
3. Tried moving firebase app initialization to attachBaseContext (before onCreate) and got a different crash (probably because a lot of the context isn't available yet)
4. Tried to disable R8 per https://issuetracker.google.com/issues/120277396

**Comment by zsweigart** [comment]
The issue is resolved if I remove the firebase performance plugin

```
apply plugin: 'com.google.firebase.firebase-perf
```

but I want the network data so I would prefer not to have to use that solution

**Comment by lurecas** [comment]
It has started to happen widly to me, but the different thing from @zsweigart is that it has only happened on certain Samsung devices :
![image](https://user-images.githubusercontent.com/4138011/52353568-1ffcc700-2a2f-11e9-819c-af3d1d5ce940.png)

**Comment by Archinamon** [comment]
We're facing this issue mostly on Xiaomi devices. We have several processes within our app. One of them handles background reporting of analytics grabbing within the app. Sometimes when this analytics process tries to execute network operation it crashes with `Make sure to call FirebaseApp.initializeApp(Context) first` exception.

![image](https://user-images.githubusercontent.com/6616695/53722653-196b3f00-3e77-11e9-86c4-dc8fcb2de470.png)

**Comment by aguatno** [comment]
Hi @zsweigart sorry for my late response here. I found an [internal bug](https://issuetracker.google.com/119442723) for this, and it looks like the issue has been fixed in version 16.2.4 of Firebase Performance for Android.

I am going to close this issue. If you're still experiencing this problem and want to continue the discussion just leave a comment here and we are happy to re-open this.
